pagespeed/envoy/envoy_fetch.cc

```diff
@@ -0,0 +1,151 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "envoy_fetch.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <list>
+#include <map>
+#include <set>
+
+#include "net/instaweb/http/public/async_fetch.h"
+#include "net/instaweb/http/public/inflating_fetch.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/condvar.h"
+#include "pagespeed/kernel/base/message_handler.h"
+#include "pagespeed/kernel/base/pool.h"
+#include "pagespeed/kernel/base/pool_element.h"
+#include "pagespeed/kernel/base/statistics.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/kernel/base/thread_system.h"
+#include "pagespeed/kernel/base/timer.h"
+#include "pagespeed/kernel/base/writer.h"
+#include "pagespeed/kernel/http/request_headers.h"
+#include "pagespeed/kernel/http/response_headers.h"
+#include "pagespeed/kernel/http/response_headers_parser.h"
+#include "pagespeed_remote_data_fetcher.h"
+#include "pagespeed/envoy/header_utils.h"
+
+namespace net_instaweb {
+
+// Default keepalive 60s.
+const int64 keepalive_timeout_ms = 60000;
+
+PagespeedDataFetcherCallback::PagespeedDataFetcherCallback(EnvoyFetch* fetch) { 
+  fetch_ = fetch;
+}
+
+void PagespeedDataFetcherCallback::onSuccess(Envoy::Http::MessagePtr& response) {
+  fetch_->setResponse(response->headers(), response->body());
+}
+
+void PagespeedDataFetcherCallback::onFailure(FailureReason reason) {
+  std::cout << "PagespeedDataFetcherCallback::onFailure\n";
+  std::cout.flush();
+}
+
+EnvoyFetch::EnvoyFetch(const GoogleString& url,
+                   AsyncFetch* async_fetch,
+                   MessageHandler* message_handler,
+                   EnvoyClusterManager& cluster_manager)
+    : str_url_(url),
+      fetcher_(NULL),
+      async_fetch_(async_fetch),
+      message_handler_(message_handler),
+      cluster_manager_(cluster_manager),
+      done_(false),
+      content_length_(-1),
+      content_length_known_(false) {
+}
+
+EnvoyFetch::~EnvoyFetch() {
+}
+
+void EnvoyFetch::FetchWithEnvoy() {
+  envoy::api::v2::core::HttpUri http_uri;
+  http_uri.set_uri("http://localhost:80");
+  http_uri.set_cluster("cluster1");
+  std::string uriHash("123456789");
+
+  cb_ptr_ = std::make_unique<PagespeedDataFetcherCallback>(this);
+  std::unique_ptr<PagespeedRemoteDataFetcher> PagespeedRemoteDataFetcherPtr = std::make_unique<PagespeedRemoteDataFetcher>(
+    cluster_manager_.getClusterManager(), http_uri, uriHash, *cb_ptr_);
+
+  PagespeedRemoteDataFetcherPtr->fetch();
+  cluster_manager_.getDispatcher()->run(Envoy::Event::Dispatcher::RunType::Block);
+}
+
+
+// This function is called by EnvoyUrlAsyncFetcher::StartFetch.
+void EnvoyFetch::Start() {
+  std::function<void()> fetch_fun_ptr =
+      std::bind(&EnvoyFetch::FetchWithEnvoy, this);
+  cluster_manager_.getDispatcher()->post(fetch_fun_ptr);
+  cluster_manager_.getDispatcher()->run(Envoy::Event::Dispatcher::RunType::NonBlock);
```

Comment:
This looks a bit off, I assume we're posting to an up and running event loop, I wonder why we need to call run on that here?

Reply:
Yeah I think we should just remove this line. `post()` provides a thread-safe way to post work to a dispatcher running on another thread. But when you call run here, that makes the dispatcher run on this very same thread. That does not make sense to me, and I suspect no good will come of it; is this an optimization of sorts?